### PR TITLE
Add code tabs to Other changes page

### LINKED ISSUE #2481: Update code examples to show both Scala 2 and Scala 3 syntax
We would like to have a single website documenting both Scala 2 and Scala 3 since, in most of the cases, their syntax is very similar, and their features are almost all the same. However, there can still be some syntactic differences between idiomatic Scala 2 code and idiomatic Scala 3 code. Therefore, we want to display both styles in our code examples by using tabs:

Screenshot 2022-07-27 at 14-37-55 Futures and Promises

For code examples that are idiomatic in both Scala 2 and Scala 3, we show a single tab labelled “Scala 2 and 3”:

Screenshot from 2022-07-27 14-39-09

---

Unfortunately, there are lots of code examples in the documentation, and updating all of them is a huge (and repetitive) task for one person. But if a few dozen members of the community were each updating a couple of pages, the job would be done very quickly. So, everyone is welcome to help!

The policy for updating code examples is the following: show the most idiomatic style for all the code examples, for Scala 2 and Scala 3. For Scala 3 code examples, this means using the [new control syntax](https://docs.scala-lang.org/scala3/reference/other-new-features/control-syntax.html), the [indentation-significant syntax](https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html), `using` and `given`, `@main`, `.*` wildcard import, `?` wildcard type argument, tuple indexed access, etc.

You can find the instructions to write tabbed code examples in the [documentation](https://docs.scala-lang.org/contribute/add-guides.html#code-blocks).

Here is the list of pages that need to be updated. If you are interested in contributing, please add a comment to announce that you start working on a page, and I will mark it “in progress” to prevent overlapping contributions:

- [x] Tour of Scala (source files are in the `_tour/` directory)
  - [x] ~Basics~ #2458
  - [x] ~Unified Types~ #2492
  - [x] ~Classes~ #2474
  - [x] ~Default Parameter Values~ #2493
  - [x] ~Named Arguments~ #2521
  - [x] ~Traits~ #2455
  - [x] ~Tuples~ #2476
  - [x] ~Class Composition with Mixins~ #2524
  - [x] ~Higher-order Functions~ #2525
  - [x] ~Nested Methods~ #2528
  - [x] ~Multiple Parameter Lists (Currying)~ #2530
  - [x] ~Case Classes~ https://github.com/scala/docs.scala-lang/pull/2533
  - [x] ~Pattern Matching~ #2461
  - [x] ~Singleton Objects~ #2497
  - [x] ~Regular Expression Patterns~ #2534
  - [x] ~Extractor Objects~ #2535
  - [x] ~For Comprehensions~ #2494
  - [x] ~Generic Classes~ #2537
  - [x] ~Variances~ #2547
  - [x] ~Upper Type Bounds~ #2548
  - [x] ~Lower Type Bounds~ #2549
  - [x] ~Inner Classes~ #2550
  - [x] ~Abstract Type Members~ #2551
  - [x] ~Compound Types~ #2552
  - [x] ~Self-type~ #2553
  - [x] ~Implicit Parameters~ #2496
  - [x] ~Implicit Conversions~ #2630
  - [x] ~Polymorphic Methods~ #2554
  - [x] ~Type Inference~ #2557
  - [x] ~Operators~ #2561
  - [x] ~By-name Parameters~ #2564
  - [x] ~Annotations~ #2566
  - [x] ~Packages and Imports~ #2570
  - [x] ~Package Objects~ #2579
- [ ] Guides and Overviews (source files are in the `_overviews/` directory)
  - [x] Scala Collections
    - [x] ~Introduction~ https://github.com/scala/docs.scala-lang/pull/2571
    - [x] ~Mutable and Immutable Collections~ https://github.com/scala/docs.scala-lang/pull/2571
    - [x] ~Trait Iterable~ https://github.com/scala/docs.scala-lang/pull/2571
    - [x] ~The sequence traits Seq, IndexedSeq, and LinearSeq~ https://github.com/scala/docs.scala-lang/pull/2571
    - [x] ~Sets~ https://github.com/scala/docs.scala-lang/pull/2572
    - [x] ~Maps~ #2455
    - [x] ~Concrete Immutable Collection Classes~ https://github.com/scala/docs.scala-lang/pull/2572
    - [x] ~Concrete Mutable Collection Classes~ https://github.com/scala/docs.scala-lang/pull/2572
    - [x] ~Arrays~ https://github.com/scala/docs.scala-lang/pull/2574
    - [x] ~Strings~ https://github.com/scala/docs.scala-lang/pull/2574
    - [x] ~Performance Characteristics~ (no code examples)
    - [x] ~Equality~ https://github.com/scala/docs.scala-lang/pull/2574
    - [x] ~Views~ https://github.com/scala/docs.scala-lang/pull/2574
    - [x] ~Iterators~ https://github.com/scala/docs.scala-lang/pull/2575
    - [x] ~Creating Collections From Scratch~ https://github.com/scala/docs.scala-lang/pull/2575
    - [x] ~Conversions Between Java and Scala Collections~ https://github.com/scala/docs.scala-lang/pull/2575
  - [x] ~The Architecture of Scala Collections~ https://github.com/scala/docs.scala-lang/pull/2576#pullrequestreview-1129904740
  - [x] ~Implementing Custom Collections~ #2585
  - [x] ~Adding Custom Collection Operations~ #2589
  - [ ] Migration from Scala 2 to Scala 3
    - [x] ~Migration Tutorial~ #2597
    - [x] ~Scala 3 Syntax Rewriting~ #2643
    - [ ] Incompatibility Table
      - [x] ~Syntactic Changes~ #2722
      - [x] Dropped Features #2723
      - [x] Contextual Abstractions #2724
      - [x] Other Changed Features #2725
      - [x] ~Type Checker~ #2726
      - [ ] Type Inference
    - [ ] Compiler Plugins
      - [ ] Kind Projector Migration
  - [ ] Scala 3 Macros
  - [ ] Value Classes and Universal Traits
  - [ ] An Overview of TASTy
  - [x] ~String Interpolation~ #2455
  - [ ] Implicit Classes
  - [ ] Library Author Guide
  - [x] ~Futures and Promises~ #2467
  - [ ] Optimizer
- [ ] Scala 3 Book (source files are in the `_overviews/scala3-book` directory)
  - [x] ~Scala 3 Features~ #2495
  - [x] ~Why Scala 3?~ #2510
  - [x] A Taste of Scala
    - [x] ~Hello, World!~ #2450
    - [x] ~The REPL~ #2487
    - [x] ~Variables and Data Types~ #2568
    - [x] ~Control Structures~ #2580
    - [x] ~Domain Modeling~ #2581
    - [x] ~Methods~ #2582
    - [x] ~First-Class Functions~ #2583
    - [x] ~Singleton Objects~ #2584
    - [x] ~Collections~ #2586
    - [x] ~Contextual Abstractions~ #2587
    - [x] ~Top-level Definitions~ #2588
  - [x] ~A First Look at Types~ #2511
  - [x] ~Control Structures~ #2461
  - [x] ~Domain Modeling~ (no code)
    - [x] ~Tools~ #2590
    - [x] ~OOP Modeling~ #2591
    - [x] ~FP Modeling~ #2592 
  - [x] ~Methods~ (no code)
    - [x] ~Method Features~ #2594
    - [x] ~main Methods~ #2609
  - [ ] Functions
    - [x] ~Anonymous Functions~ #2602
    - [x] ~Function Variables~ #2623
    - [ ] Eta Expansion
    - [x] ~Higher-Order Functions~ #2598
    - [x] ~Write Your Own map Method~ #2625
    - [x] ~Creating a Method That Returns a Function~ #2664 
  - [x] Packaging and Imports (@anatoliykmetyuk)
  - [x] ~Scala Collections~ #2614
    - [x] ~Collections Types~ #2614
    - [x] ~Collections Methods~ #2614
  - [x] ~Functional Programming~ #2617
    - [x] ~What is Functional Programming~ (no code)
    - [x] ~Immutable Values~ #2617
    - [x] ~Pure Functions~ #2617
    - [x] ~Functions Are Values~ #2617
    - [x] ~Functional Error Handling~ #2617
  - [x] ~Types and the Type System~ #2599
    - [x] ~Inferred Types~ #2599
    - [x] ~Generics~ #2599
    - [x] ~Intersection Types~ #2653
    - [ ] Union Types
    - [ ] Algebraic Data Types
    - [ ] Variance
    - [ ] Opaque Types
    - [ ] Structural Types
    - [ ] Dependent Function Types
    - [ ] Other Types
  - [ ] ~Contextual Abstractions~ (in progress)
    - [x] ~Given Instances and Using Clauses~ #2640
    - [ ] ~Type Classes~
    - [x] ~Context Bounds~ #2689
    - [x] ~Given Imports~ #2689
    - [ ] ~Extension Methods~
    - [ ] ~Implementing Type Classes~
    - [x] ~Multiversal Equality~ #2689
    - [ ] ~Implicit Conversions~
  - [ ] Concurrency
  - [ ] Scala Tools
    - [ ] Building and Testing Scala Projects with sbt
    - [ ] Worksheets
  - [ ] Interacting with Java
  - [ ] Scala for Java Developers
  - [ ] Scala for JavaScript Developers
  - [ ] Scala for Python Developers
- [ ] Tutorials (source files are in the `_overviews/tutorials` directory or `_getting-started/`)
  - [x] ~Getting started with Scala and sbt on the command line~ #2475 
  - [x] ~Getting started with Scala in IntelliJ~ #2454 
  - [x] ~Scala for Java Programmers~ #2499
  - [ ] Scala on Android
  - [ ] Scala with Maven

---

Note: in some cases, not only the code examples need to be updated, but also the text in the documentation (e.g. `@main` methods, extension methods, contextual abstractions, enums). In that case, the task is more complicated. Feel free to ask for help on the `#scala-contributors` channel of the [Scala discord server](https://discord.com/invite/scala), or consider switching to a simpler page.

### PR DESCRIPTION
**Pull request by Dedelweiss**
Here is my PR to add code tabs to the Other changes page

Fixes #2481

### CODE REVIEW

**Review by julienrf — CHANGES_REQUESTED** [review]
Hey Lucas, I started reviewing this but it seems you always put “Scala 3 Only” as a tab title for code that compiles with Scala 2 only. Or do I miss something?

**Comment by Dedelweiss** [comment]
> Hey Lucas, I started reviewing this but it seems you always put “Scala 3 Only” as a tab title for code that compiles with Scala 2 only. Or do I miss something?

Oh yes, that's a misunderstanding on my part. For me, since there was an error afterwards it was like saying that this code in scala 3 has the error right afterwards. So what I can do is put the code in scala 2 only and say that in scala 3 only it produces the error.

**Comment by julienrf** [comment]
> what I can do is put the code in scala 2 only and say that in scala 3 only it produces the error.

That sounds good to me, thank you!

**Comment by Dedelweiss** [comment]
When there is an error in the code itself like the example below, do I leave it in scala 3 only or do I put in scala 2 and 3?
Screenshot 2023-03-09 at 09 43 43

**Comment by julienrf** [comment]
Yeah, that is tricky, I understand your reasoning. By putting the “Scala 3 Only” you mean “that error that we mention in the code snippet happens in Scala 3 Only”. But everywhere else on the website we use “Scala 3 Only” to put a code snippet that is valid only in Scala 3.

I think we could make things clearer with the following approach:
- Put the “Scala 2 Only” tab to indicate that the code snippet is valid in Scala 2 only
- In the comment where we show the error message reported when we compile that snippet with Scala 3, make it explicit that the error message is specific to Scala 3:
  ~~~ scala
  pojo.setFooBar("hello") // In Scala 3, Not Found Error: value setFooBar is not a member of Pojo
  ~~~

How does that sound?

**Comment by Dedelweiss** [comment]
> How does that sound?

Oh yes, indeed, I think it is better in terms of understanding. Thank you for your attention. And so I put back all the {% highlight diff %}? I thought it was a difference between Scala 2 and Scala 3 so I put them in the tabs, but it is true that in terms of understanding it is good to have the difference with the -/+. Should I still put the {% highlight diff %} in the tabs of Scala 3 only (or others)?  Like this example below : 

Screenshot 2023-03-09 at 11 41 25

**Comment by julienrf** [comment]
In that case, since we show a diff and not a code snippet that the readers than copy-paste, we can leave out the tab.

File: _overviews/scala3-migration/incompat-other-changes.md [hunk 1]
```diff
@@ -71,22 +79,26 @@ The [Scala 3 migration compilation](tooling-migration-mode.html) can automatical
 
 The Scala 3 compiler requires the constructor of private classes to be private.
 
+{% tabs scala-2-constructor_1 %}
+{% tab 'Scala 2 Only' %}
 For instance, in the example:
```
Comment (julienrf) [inline _overviews/scala3-migration/incompat-other-changes.md:84]: This is very minor, but we should put only the code snippet in the tab, not this sentence. The only situation where we put other things in a tab is when there are multiple tabs showing different pieces of code and the text need to refer specifically to the content of each snippet.

File: _overviews/scala3-migration/incompat-other-changes.md [hunk 2]
```diff
@@ -96,9 +108,11 @@ The solution is to make the constructor private, since the class is private.
 
 In Scala 3, overriding a concrete def with an abstract def causes subclasses to consider the def abstract, whereas in Scala 2 it was considered as concrete.
 
-In the following piece of code, the `bar` method in `C` is considered concrete by the Scala 2.13 compiler but abstract by the Scala 3 compiler, causing the following error.
+{% tabs scala-2-abstract_1 %}
+{% tab 'Scala 2 Only' %}
 
-```scala
+In the following piece of code, the `bar` method in `C` is considered concrete by the Scala 2.13 compiler but abstract by the Scala 3 compiler, causing the following error.
```
Comment (julienrf) [inline _overviews/scala3-migration/incompat-other-changes.md:114]: Same comment here about that sentence that should be outside the tab.

File: _overviews/scala3-migration/incompat-other-changes.md [hunk 3]
```diff
@@ -107,45 +121,55 @@ trait B extends A {
...
 
 The companion object of a case class does not extend any of the `Function{0-23}` traits anymore.
 In particular, it does not inherit their methods: `tupled`, `curried`, `andThen`, `compose`...
 
-For instance, this is not permitted anymore:
+{% tabs scala-2-companion_1 %}
+{% tab 'Scala 2 Only' %}
 
-```scala
+For instance, this is not permitted anymore:
```
Comment (julienrf) [inline _overviews/scala3-migration/incompat-other-changes.md:141]: Here too

**Review by julienrf — CHANGES_REQUESTED** [review]
Thank you again Lucas for your help, but I’ve noticed a pattern that should be fixed: the tabs should primarily wrap code snippets, not text, unless the tabs contain different code snippets and we want to add some text specifically for each snippet.

File: _overviews/scala3-migration/incompat-other-changes.md (hunk 1 above)
Comment (Dedelweiss) [reply _overviews/scala3-migration/incompat-other-changes.md:84]: Okay, sorry for the mistake, I'm correcting it on all the other PRs.

File: _overviews/scala3-migration/incompat-other-changes.md [hunk 4]
```diff
@@ -139,13 +156,16 @@ A cross-compiling solution is to explicitly eta-expand the method `Foo.apply`.
 {% endhighlight %}
 
 Or, for performance reasons, you can introduce an intermediate function value.
-
-```scala
+{% tabs scala-3-companion_2 %}
+{% tab 'Scala 3 Only' %}
```
Comment (julienrf) [inline _overviews/scala3-migration/incompat-other-changes.md:160]: This code snippet cross-compiles.

```suggestion
{% tab 'Scala 2 and 3' %}
```

**Review by julienrf — APPROVED** [review]
Thank you for adding all the code tabs!

**Pull request merged**